time/src/main/java/net/jqwik/time/internal/properties/arbitraries/DefaultLocalTimeArbitrary.java

```diff
@@ -85,97 +82,113 @@ private long calculateLongEnd(LocalTime effectiveMin, LocalTime effectiveMax) {
+	}
+
+	private static void throwTimeAndPrecisionException(String val, boolean minimum, ChronoUnit precision) {
+		String minMax = minimum ? "minimum" : "maximum";
+		throw new IllegalArgumentException(
+			String
+				.format("Can't use %s as %s time with precision %s.%nYou may want to round the time to %s or change the precision.", val, minMax, precision, precision)
+		);
+	}
+
+	public static LocalTime effectiveMax(
+		LocalTimeBetween timeBetween,
+		HourBetween hourBetween,
+		MinuteBetween minuteBetween,
+		SecondBetween secondBetween,
+		OfPrecision ofPrecision
+	) {
+		LocalTime effective = timeBetween.getMax() != null ? timeBetween.getMax() : ofPrecision.maxPossibleLocalTime();
+		checkTimeValueAndPrecision(effective, ofPrecision, false);
 		if (hourBetween.getMax() < effective.getHour()) {
 			effective = effective.withHour(hourBetween.getMax());
-			if (minuteBetween.getMax() < effective.getMinute()) {
-				effective = effective.withMinute(minuteBetween.getMax());
-				if (secondBetween.getMax() < effective.getSecond()) {
-					effective = effective.withSecond(secondBetween.getMax());
-					effective = effective.withNano(999_999_999);
-				}
-			}
+			effective = setEffectiveMaxValues(effective, ofPrecision, MINUTES);
+		}
+		if (minuteBetween.getMax() < effective.getMinute()) {
+			effective = effective.withMinute(minuteBetween.getMax());
+			effective = setEffectiveMaxValues(effective, ofPrecision, SECONDS);
+		}
+		if (secondBetween.getMax() < effective.getSecond()) {
+			effective = effective.withSecond(secondBetween.getMax());
+			effective = setEffectiveMaxValues(effective, ofPrecision, NANOS);
 		}
-		effective = calculateEffectiveMaxWithPrecision(effective, ofPrecision);
 		return effective;
 	}
 
-	public static LocalTime calculateEffectiveMaxWithPrecision(LocalTime effective, OfPrecision ofPrecision) {
-		return calculateEffectiveMaxWithPrecision(effective, ofPrecision.get());
-	}
-
-	public static LocalTime calculateEffectiveMaxWithPrecision(LocalTime effective, ChronoUnit ofPrecision) {
-		switch (ofPrecision) {
-			case HOURS:
-				effective = effective.withMinute(0);
+	private static LocalTime setEffectiveMaxValues(LocalTime effective, OfPrecision ofPrecision, ChronoUnit precision) {
+		switch (precision) {
 			case MINUTES:
-				effective = effective.withSecond(0);
+				effective = ofPrecision.get().compareTo(MINUTES) <= 0 ? effective.withMinute(59) : effective.withMinute(0);
 			case SECONDS:
-				effective = effective.withNano(effective.getNano() % 1_000_000);
-			case MILLIS:
-				effective = effective.withNano((effective.getNano() / 1_000_000) * 1_000_000 + effective.getNano() % 1_000);
-			case MICROS:
-				effective = effective.withNano(effective.getNano() - effective.getNano() % 1_000);
+				effective = ofPrecision.get().compareTo(SECONDS) <= 0 ? effective.withSecond(59) : effective.withSecond(0);
```

Comment:
Can those comparisons be replaced by methods in OfPrecision? e.g. ofPrecision.isLessOrEqualTo(..)?